source/extensions/filters/network/metadata_exchange/metadata_exchange.cc

```diff
@@ -103,6 +103,14 @@ Network::FilterStatus MetadataExchangeFilter::onData(Buffer::Instance& data, boo
   case NeedMoreDataInitialHeader: {
     tryReadInitialProxyHeader(data);
     if (conn_state_ == NeedMoreDataInitialHeader) {
+      if (end_stream) {
+        // Upstream has entered a half-closed state, and will be sending no more data.
+        // Since this plugin would expect additional headers, but none is forthcoming,
+        // do not block the tcp_proxy downstream of us from draining the buffer.
+        ENVOY_LOG(warn, "Upstream closed early, aborting istio-peer-exchange");
```

Comment:
This should be debug or trace level.

Reply:
Please fix log level to be less noisy - will remove the label then.

Reply:
Yes, I believe the difference is due to the 0-RTT behavior of TLS 1.3 changing the timing of the this extension relative to the handshake. Specifically, with 1.2, mx is never invoked because the upstream always closes early with a handshake failure.
Regarding the log level, I can certainly change that and will do so. My rationale had been that this failure is rare and would happen only when one does not define something like a startupProbe in the pod spec, and so users that debugged this may have been more likely to see the warning were it to happen, as it did for us.

Reply:
The log level is driven by Envoy conventions, which usually limits per-connection and per-request lines to be <= debug, to avoid overwhelming the log service.

If I understand correctly, this issue can happen with any mis-configured mesh using TLS 1.3. Any client sidecar connection that is accepted and then half-closed would cause a connection leak from the app to the client sidecar. This may happen when trying to use mTLS to external servers (out-of-mesh). If it's truly 0-RTT timing then the server doesn't even have to be TLS to cause an issue.

Reply:
Misconfigured external servers may trigger this issue, although I believe we only get to the problematic point in the state machine if the server advertises the correct magic string in the ALPN sent back in the first packet.
Thank you for the review!